<img src='Data/movie.png'>

In [67]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

# Loading the Data

In [68]:
df_2000 = pd.read_csv('/Users/christianrim/Documents/IMDB Movies/final_tmdb_data_2000.csv',lineterminator='\n')
df_2001 = pd.read_csv('/Users/christianrim/Documents/IMDB Movies/final_tmdb_data_2001.csv',lineterminator='\n')

In [69]:
# display the 2000 movie dataset
df_2000.head()
df_2001.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6153 entries, 0 to 6152
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                6153 non-null   object 
 1   adult                  6152 non-null   float64
 2   backdrop_path          1782 non-null   object 
 3   belongs_to_collection  265 non-null    object 
 4   budget                 6152 non-null   float64
 5   genres                 6152 non-null   object 
 6   homepage               391 non-null    object 
 7   id                     6152 non-null   float64
 8   original_language      6152 non-null   object 
 9   original_title         6152 non-null   object 
 10  overview               5409 non-null   object 
 11  popularity             6152 non-null   float64
 12  poster_path            4591 non-null   object 
 13  production_companies   6152 non-null   object 
 14  production_countries   6152 non-null   object 
 15  rele

In [70]:
df =  pd.concat([df_2000,df_2001])
df.head()

imdb_id  adult                     backdrop_path belongs_to_collection  \
0          0    NaN                               NaN                   NaN   
1  tt0015414   0.00                               NaN                   NaN   
2  tt0113026   0.00  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN   
3  tt0113086   0.00                               NaN                   NaN   
4  tt0113092   0.00                               NaN                   NaN   

       budget                                             genres homepage  \
0         NaN                                                NaN      NaN   
1        0.00                      [{'id': 18, 'name': 'Drama'}]      NaN   
2 10000000.00  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...      NaN   
3        0.00                                                 []      NaN   
4        0.00           [{'id': 878, 'name': 'Science Fiction'}]      NaN   

         id original_language          original_title  ... revenue  runtime  \
0       NaN               NaN                     NaN  ...     NaN      NaN   
1 607290.00                es  La tierra de los toros  ...    0.00    53.00   
2  62127.00                en         The Fantasticks  ...    0.00    86.00   
3 612666.00                en  Florentino y el diablo  ...    0.00    90.00   
4 110977.00                en           For the Cause  ...    0.00   100.00   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'No Language', 'iso_639_1': ...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3                                                 []  Released   
4  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                         tagline                   title  \
0                                            NaN                     NaN   
1                                            NaN   The Land of the Bulls   
2  Try to remember the first time magic happened         The Fantasticks   
3                                            NaN  Florentino y el diablo   
4   The ultimate showdown on a forbidden planet.           For the Cause   

   video  vote_average vote_count certification  
0    NaN           NaN        NaN           NaN  
1   0.00          0.00       0.00           NaN  
2   0.00          5.50      22.00           NaN  
3   0.00          5.00       1.00           NaN  
4   0.00          5.10       8.00           NaN  

[5 rows x 26 columns]

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11978 entries, 0 to 6152
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                11978 non-null  object 
 1   adult                  11976 non-null  float64
 2   backdrop_path          3456 non-null   object 
 3   belongs_to_collection  519 non-null    object 
 4   budget                 11976 non-null  float64
 5   genres                 11976 non-null  object 
 6   homepage               701 non-null    object 
 7   id                     11976 non-null  float64
 8   original_language      11976 non-null  object 
 9   original_title         11976 non-null  object 
 10  overview               10467 non-null  object 
 11  popularity             11976 non-null  float64
 12  poster_path            8972 non-null   object 
 13  production_companies   11976 non-null  object 
 14  production_countries   11976 non-null  object 
 15  rel

# Cleaning the data

In [72]:
# Removing columns with a lot of missing values
df = df.drop(['belongs_to_collection','backdrop_path','poster_path','tagline','homepage'], axis=1)
df.head()

imdb_id  adult      budget  \
0          0    NaN         NaN   
1  tt0015414   0.00        0.00   
2  tt0113026   0.00 10000000.00   
3  tt0113086   0.00        0.00   
4  tt0113092   0.00        0.00   

                                              genres        id  \
0                                                NaN       NaN   
1                      [{'id': 18, 'name': 'Drama'}] 607290.00   
2  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...  62127.00   
3                                                 [] 612666.00   
4           [{'id': 878, 'name': 'Science Fiction'}] 110977.00   

  original_language          original_title  \
0               NaN                     NaN   
1                es  La tierra de los toros   
2                en         The Fantasticks   
3                en  Florentino y el diablo   
4                en           For the Cause   

                                            overview  popularity  \
0                                                NaN         NaN   
1                                                NaN        0.60   
2  Two rural teens sing and dance their way throu...        3.08   
3  Florentino is a young horseman and singer, a f...        0.60   
4  Earth is in a state of constant war and two co...        2.10   

                                production_companies  ... release_date  \
0                                                NaN  ...          NaN   
1  [{'id': 77948, 'logo_path': None, 'name': 'Mus...  ...   1924-01-01   
2  [{'id': 51207, 'logo_path': None, 'name': 'Sul...  ...   2000-09-22   
3                                                 []  ...   2000-04-29   
4  [{'id': 7405, 'logo_path': '/rfnws0uY8rsNAsrLb...  ...   2000-11-15   

  revenue  runtime                                   spoken_languages  \
0     NaN      NaN                                                NaN   
1    0.00    53.00  [{'english_name': 'No Language', 'iso_639_1': ...   
2    0.00    86.00  [{'english_name': 'English', 'iso_639_1': 'en'...   
3    0.00    90.00                                                 []   
4    0.00   100.00  [{'english_name': 'English', 'iso_639_1': 'en'...   

     status                   title video  vote_average  vote_count  \
0       NaN                     NaN   NaN           NaN         NaN   
1  Released   The Land of the Bulls  0.00          0.00        0.00   
2  Released         The Fantasticks  0.00          5.50       22.00   
3  Released  Florentino y el diablo  0.00          5.00        1.00   
4  Released           For the Cause  0.00          5.10        8.00   

   certification  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

[5 rows x 21 columns]

In [73]:
df['budget'].describe()

count       11976.00
mean      1146148.83
std       8177761.87
min             0.00
25%             0.00
50%             0.00
75%             0.00
max     140000000.00
Name: budget, dtype: float64

In [74]:
# getting rid of the 0's in the 'budget' and 'revenue' column
filtered_df = df[(df['budget'] != 0) & (df['revenue'] != 0)]

In [75]:
filtered_df.head()

imdb_id  adult      budget  \
0           0    NaN         NaN   
7   tt0118694   0.00   150000.00   
19  tt0120467   0.00   120000.00   
22  tt0120630   0.00 45000000.00   
25  tt0120753   0.00  8000000.00   

                                               genres       id  \
0                                                 NaN      NaN   
7   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   843.00   
19  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name... 19085.00   
22  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  7443.00   
25  [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...   318.00   

   original_language            original_title  \
0                NaN                       NaN   
7                 cn                      花樣年華   
19                en                    Vulgar   
22                en               Chicken Run   
25                en  The Million Dollar Hotel   

                                             overview  popularity  \
0                                                 NaN         NaN   
7   Two neighbors become intimate after discoverin...       23.46   
19  Vulgar is about a man who is a children's clow...        4.57   
22  The creators of Wallace & Gromit bring you an ...       38.20   
25  The Million Dollar Hotel starts with a jump fr...       11.32   

                                 production_companies  ... release_date  \
0                                                 NaN  ...          NaN   
7   [{'id': 539, 'logo_path': '/iPLtePguIzOPNtAWfT...  ...   2000-09-29   
19  [{'id': 102045, 'logo_path': None, 'name': 'Ch...  ...   2002-04-26   
22  [{'id': 521, 'logo_path': '/kP7t6RwGz2AvvTkvnI...  ...   2000-06-23   
25  [{'id': 1490, 'logo_path': None, 'name': 'Kint...  ...   2000-02-09   

        revenue  runtime                                   spoken_languages  \
0           NaN      NaN                                                NaN   
7   12854953.00    99.00  [{'english_name': 'Cantonese', 'iso_639_1': 'c...   
19     14904.00    87.00  [{'english_name': 'English', 'iso_639_1': 'en'...   
22 224834564.00    84.00  [{'english_name': 'English', 'iso_639_1': 'en'...   
25    105983.00   122.00  [{'english_name': 'English', 'iso_639_1': 'en'...   

      status                     title video  vote_average  vote_count  \
0        NaN                       NaN   NaN           NaN         NaN   
7   Released      In the Mood for Love  0.00          8.11     2163.00   
19  Released                    Vulgar  0.00          5.50       44.00   
22  Released               Chicken Run  0.00          6.70     4313.00   
25  Released  The Million Dollar Hotel  0.00          5.92      274.00   

    certification  
0             NaN  
7              PG  
19              R  
22              G  
25              R  

[5 rows x 21 columns]

In [76]:
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 404 entries, 0 to 5933
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   imdb_id               404 non-null    object 
 1   adult                 402 non-null    float64
 2   budget                402 non-null    float64
 3   genres                402 non-null    object 
 4   id                    402 non-null    float64
 5   original_language     402 non-null    object 
 6   original_title        402 non-null    object 
 7   overview              401 non-null    object 
 8   popularity            402 non-null    float64
 9   production_companies  402 non-null    object 
 10  production_countries  402 non-null    object 
 11  release_date          401 non-null    object 
 12  revenue               402 non-null    float64
 13  runtime               402 non-null    float64
 14  spoken_languages      402 non-null    object 
 15  status                

## How many movies had at least some valid financial information 

In [77]:
# Select desired columns
selected_cols = ['title', 'revenue', 'budget']
result_df = filtered_df.loc[:, selected_cols]
result_df[:10]

title      revenue       budget
0                        NaN          NaN          NaN
7       In the Mood for Love  12854953.00    150000.00
19                    Vulgar     14904.00    120000.00
22               Chicken Run 224834564.00  45000000.00
25  The Million Dollar Hotel    105983.00   8000000.00
26    Mission: Impossible II 546388105.00 125000000.00
27                     X-Men 296339527.00  75000000.00
28                Titan A.E.  36754634.00  75000000.00
29  The Emperor's New Groove 169327687.00 100000000.00
33              Return to Me  36609995.00  24000000.00

In [78]:
tmdb_data = df[(df['revenue'] > 0) & (df['budget'] > 0)]
# Group the filtered DataFrame by 'movie' and get the count of movies in each group
count_by_movie = tmdb_data.groupby('title').size()

# Display the result
print(count_by_movie)

title
102 Dalmatians            1
15 Minutes                1
2001: A Space Travesty    1
28 Days                   1
3 Strikes                 1
                         ..
Xchange                   1
Y Tu Mamá También         1
Yajamana                  1
Yamakasi                  1
Zoolander                 1
Length: 402, dtype: int64


- There were 402 movies with at least one financial data. 

In [79]:
tmdb_data = tmdb_data.drop(columns=['production_companies','spoken_languages'], axis=1)
tmdb_data.head()

imdb_id  adult       budget  \
7   tt0118694   0.00    150000.00   
19  tt0120467   0.00    120000.00   
22  tt0120630   0.00  45000000.00   
25  tt0120753   0.00   8000000.00   
26  tt0120755   0.00 125000000.00   

                                               genres       id  \
7   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   843.00   
19  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name... 19085.00   
22  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  7443.00   
25  [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...   318.00   
26  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   955.00   

   original_language            original_title  \
7                 cn                      花樣年華   
19                en                    Vulgar   
22                en               Chicken Run   
25                en  The Million Dollar Hotel   
26                en    Mission: Impossible II   

                                             overview  popularity  \
7   Two neighbors become intimate after discoverin...       23.46   
19  Vulgar is about a man who is a children's clow...        4.57   
22  The creators of Wallace & Gromit bring you an ...       38.20   
25  The Million Dollar Hotel starts with a jump fr...       11.32   
26  With computer genius Luther Stickell at his si...       33.85   

                                 production_countries release_date  \
7   [{'iso_3166_1': 'FR', 'name': 'France'}, {'iso...   2000-09-29   
19  [{'iso_3166_1': 'US', 'name': 'United States o...   2002-04-26   
22  [{'iso_3166_1': 'GB', 'name': 'United Kingdom'...   2000-06-23   
25  [{'iso_3166_1': 'US', 'name': 'United States o...   2000-02-09   
26  [{'iso_3166_1': 'US', 'name': 'United States o...   2000-05-24   

        revenue  runtime    status                     title  video  \
7   12854953.00    99.00  Released      In the Mood for Love   0.00   
19     14904.00    87.00  Released                    Vulgar   0.00   
22 224834564.00    84.00  Released               Chicken Run   0.00   
25    105983.00   122.00  Released  The Million Dollar Hotel   0.00   
26 546388105.00   123.00  Released    Mission: Impossible II   0.00   

    vote_average  vote_count certification  
7           8.11     2163.00            PG  
19          5.50       44.00             R  
22          6.70     4313.00             G  
25          5.92      274.00             R  
26          6.11     5706.00         PG-13

In [80]:
tmdb_data = tmdb_data.drop(columns=['production_countries'], axis=1)


## How many movies are there in each of the certification categories (G/PG/PG-13/R)?

In [81]:
# counting the movie count for each certification category 
grouped_movies = tmdb_data.groupby(['certification'])
count_movies = grouped_movies.size() 
print(count_movies) 
print(count_movies.sum())

certification
G         18
NR         8
PG        31
PG-13    117
R        144
dtype: int64
318


- There are 318 movies for each certification categories. 

## What is the average revenue per certification category?

In [82]:
# Set the display float format option to a non-scientific notation format
pd.set_option('display.float_format', lambda x: '%.2f' % x)

# Group the data by "certification" and calculate the average revenue for each group
avg_revenue_by_certification = tmdb_data.groupby("certification")["revenue"].mean()
avg_revenue_by_certification

certification
G        97342643.33
NR       19369745.62
PG      125078003.97
PG-13   111073959.08
R        51735537.01
Name: revenue, dtype: float64

## What is the average budget per certification category?

In [83]:
# Set the display float format option to a non-scientific notation format
pd.set_option('display.float_format', lambda x: '%.2f' % x)

# Group the data by "certification" and calculate the average revenue for each group
avg_budget_by_certification = tmdb_data.groupby("certification")["budget"].mean()
avg_budget_by_certification


certification
G       33777635.28
NR      11837750.00
PG      50362814.00
PG-13   46644422.11
R       26732737.15
Name: budget, dtype: float64

# Getting the Data to connect to MySQL Workbench

In [84]:
# the TMDB API Results 
tmdb_data.info()
tmdb_data.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 402 entries, 7 to 5933
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   imdb_id            402 non-null    object 
 1   adult              402 non-null    float64
 2   budget             402 non-null    float64
 3   genres             402 non-null    object 
 4   id                 402 non-null    float64
 5   original_language  402 non-null    object 
 6   original_title     402 non-null    object 
 7   overview           401 non-null    object 
 8   popularity         402 non-null    float64
 9   release_date       401 non-null    object 
 10  revenue            402 non-null    float64
 11  runtime            402 non-null    float64
 12  status             402 non-null    object 
 13  title              402 non-null    object 
 14  video              402 non-null    float64
 15  vote_average       402 non-null    float64
 16  vote_count         402 no

imdb_id  adult       budget  \
7   tt0118694   0.00    150000.00   
19  tt0120467   0.00    120000.00   
22  tt0120630   0.00  45000000.00   
25  tt0120753   0.00   8000000.00   
26  tt0120755   0.00 125000000.00   

                                               genres       id  \
7   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   843.00   
19  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name... 19085.00   
22  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...  7443.00   
25  [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...   318.00   
26  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   955.00   

   original_language            original_title  \
7                 cn                      花樣年華   
19                en                    Vulgar   
22                en               Chicken Run   
25                en  The Million Dollar Hotel   
26                en    Mission: Impossible II   

                                             overview  popularity  \
7   Two neighbors become intimate after discoverin...       23.46   
19  Vulgar is about a man who is a children's clow...        4.57   
22  The creators of Wallace & Gromit bring you an ...       38.20   
25  The Million Dollar Hotel starts with a jump fr...       11.32   
26  With computer genius Luther Stickell at his si...       33.85   

   release_date      revenue  runtime    status                     title  \
7    2000-09-29  12854953.00    99.00  Released      In the Mood for Love   
19   2002-04-26     14904.00    87.00  Released                    Vulgar   
22   2000-06-23 224834564.00    84.00  Released               Chicken Run   
25   2000-02-09    105983.00   122.00  Released  The Million Dollar Hotel   
26   2000-05-24 546388105.00   123.00  Released    Mission: Impossible II   

    video  vote_average  vote_count certification  
7    0.00          8.11     2163.00            PG  
19   0.00          5.50       44.00             R  
22   0.00          6.70     4313.00             G  
25   0.00          5.92      274.00             R  
26   0.00          6.11     5706.00         PG-13

In [85]:
# importing the data from 'title.basics.tsv.gz'
basics = pd.read_csv("/Users/christianrim/Documents/IMDB Movies/IMDB-Movies/Data/title_basics.csv.gz",low_memory = False)
basics.head()

tconst titleType            primaryTitle           originalTitle  \
0  tt0000001     short              Carmencita              Carmencita   
1  tt0000002     short  Le clown et ses chiens  Le clown et ses chiens   
2  tt0000003     short          Pauvre Pierrot          Pauvre Pierrot   
3  tt0000004     short             Un bon bock             Un bon bock   
4  tt0000005     short        Blacksmith Scene        Blacksmith Scene   

   isAdult  startYear  endYear  runtimeMinutes                    genres  
0        0    1894.00      NaN            1.00         Documentary,Short  
1        0    1892.00      NaN            5.00           Animation,Short  
2        0    1892.00      NaN            4.00  Animation,Comedy,Romance  
3        0    1892.00      NaN           12.00           Animation,Short  
4        0    1893.00      NaN            1.00              Comedy,Short

In [86]:
# Open saved file and preview again
ratings = pd.read_csv("/Users/christianrim/Documents/IMDB Movies/IMDB-Movies/Data/title_ratings.csv.gz", low_memory = False)
ratings.head()

tconst  averageRating  numVotes
0  tt0000001           5.70      1966
1  tt0000002           5.80       263
2  tt0000005           6.20      2604
3  tt0000006           5.10       178
4  tt0000007           5.40       817

In [87]:
# creating a column displaying a list of genres
basics['genres_split'] = basics['genres'].str.split(',')
basics

tconst  titleType               primaryTitle  \
0        tt0000001      short                 Carmencita   
1        tt0000002      short     Le clown et ses chiens   
2        tt0000003      short             Pauvre Pierrot   
3        tt0000004      short                Un bon bock   
4        tt0000005      short           Blacksmith Scene   
...            ...        ...                        ...   
9737914  tt9916848  tvEpisode              Episode #3.17   
9737915  tt9916850  tvEpisode              Episode #3.19   
9737916  tt9916852  tvEpisode              Episode #3.20   
9737917  tt9916856      short                   The Wind   
9737918  tt9916880  tvEpisode  Horrid Henry Knows It All   

                     originalTitle  isAdult  startYear  endYear  \
0                       Carmencita        0    1894.00      NaN   
1           Le clown et ses chiens        0    1892.00      NaN   
2                   Pauvre Pierrot        0    1892.00      NaN   
3                      Un bon bock        0    1892.00      NaN   
4                 Blacksmith Scene        0    1893.00      NaN   
...                            ...      ...        ...      ...   
9737914              Episode #3.17        0    2010.00      NaN   
9737915              Episode #3.19        0    2010.00      NaN   
9737916              Episode #3.20        0    2010.00      NaN   
9737917                   The Wind        0    2015.00      NaN   
9737918  Horrid Henry Knows It All        0    2014.00      NaN   

         runtimeMinutes                      genres  \
0                  1.00           Documentary,Short   
1                  5.00             Animation,Short   
2                  4.00    Animation,Comedy,Romance   
3                 12.00             Animation,Short   
4                  1.00                Comedy,Short   
...                 ...                         ...   
9737914             NaN         Action,Drama,Family   
9737915             NaN         Action,Drama,Family   
9737916             NaN         Action,Drama,Family   
9737917           27.00                       Short   
9737918           10.00  Adventure,Animation,Comedy   

                           genres_split  
0                  [Documentary, Short]  
1                    [Animation, Short]  
2          [Animation, Comedy, Romance]  
3                    [Animation, Short]  
4                       [Comedy, Short]  
...                                 ...  
9737914         [Action, Drama, Family]  
9737915         [Action, Drama, Family]  
9737916         [Action, Drama, Family]  
9737917                         [Short]  
9737918  [Adventure, Animation, Comedy]  

[9737919 rows x 10 columns]

In [88]:
# using the explode method to separate the list into new rows
title_basics = basics.explode('genres_split')
title_basics

tconst  titleType               primaryTitle  \
0        tt0000001      short                 Carmencita   
0        tt0000001      short                 Carmencita   
1        tt0000002      short     Le clown et ses chiens   
1        tt0000002      short     Le clown et ses chiens   
2        tt0000003      short             Pauvre Pierrot   
...            ...        ...                        ...   
9737916  tt9916852  tvEpisode              Episode #3.20   
9737917  tt9916856      short                   The Wind   
9737918  tt9916880  tvEpisode  Horrid Henry Knows It All   
9737918  tt9916880  tvEpisode  Horrid Henry Knows It All   
9737918  tt9916880  tvEpisode  Horrid Henry Knows It All   

                     originalTitle  isAdult  startYear  endYear  \
0                       Carmencita        0    1894.00      NaN   
0                       Carmencita        0    1894.00      NaN   
1           Le clown et ses chiens        0    1892.00      NaN   
1           Le clown et ses chiens        0    1892.00      NaN   
2                   Pauvre Pierrot        0    1892.00      NaN   
...                            ...      ...        ...      ...   
9737916              Episode #3.20        0    2010.00      NaN   
9737917                   The Wind        0    2015.00      NaN   
9737918  Horrid Henry Knows It All        0    2014.00      NaN   
9737918  Horrid Henry Knows It All        0    2014.00      NaN   
9737918  Horrid Henry Knows It All        0    2014.00      NaN   

         runtimeMinutes                      genres genres_split  
0                  1.00           Documentary,Short  Documentary  
0                  1.00           Documentary,Short        Short  
1                  5.00             Animation,Short    Animation  
1                  5.00             Animation,Short        Short  
2                  4.00    Animation,Comedy,Romance    Animation  
...                 ...                         ...          ...  
9737916             NaN         Action,Drama,Family       Family  
9737917           27.00                       Short        Short  
9737918           10.00  Adventure,Animation,Comedy    Adventure  
9737918           10.00  Adventure,Animation,Comedy    Animation  
9737918           10.00  Adventure,Animation,Comedy       Comedy  

[15507172 rows x 10 columns]

In [89]:
# exploding the series using explode() and take the unqiue() entries
unique_genres = basics['genres_split'].explode().unique()
unique_genres

array(['Documentary', 'Short', 'Animation', 'Comedy', 'Romance', 'Sport',
       'News', 'Drama', 'Fantasy', 'Horror', 'Biography', 'Music', 'War',
       'Crime', 'Western', 'Family', 'Adventure', 'Action', 'History',
       'Mystery', nan, 'Sci-Fi', 'Musical', 'Thriller', 'Film-Noir',
       'Talk-Show', 'Game-Show', 'Reality-TV', 'Adult'], dtype=object)

In [90]:
title_basics.head(3)

tconst titleType            primaryTitle           originalTitle  \
0  tt0000001     short              Carmencita              Carmencita   
0  tt0000001     short              Carmencita              Carmencita   
1  tt0000002     short  Le clown et ses chiens  Le clown et ses chiens   

   isAdult  startYear  endYear  runtimeMinutes             genres genres_split  
0        0    1894.00      NaN            1.00  Documentary,Short  Documentary  
0        0    1894.00      NaN            1.00  Documentary,Short        Short  
1        0    1892.00      NaN            5.00    Animation,Short    Animation

### Creating 'Title Genres' Table

In [91]:
# creating a new table for 'title genres' 
title_genres = title_basics[['tconst','genres_split']].copy()
title_genres.head()

tconst genres_split
0  tt0000001  Documentary
0  tt0000001        Short
1  tt0000002    Animation
1  tt0000002        Short
2  tt0000003    Animation

In [92]:
# making a genre mapper dictionary
genre_ints = range(len(unique_genres))
genre_map = dict(zip(unique_genres,genre_ints))
genre_map

{'Documentary': 0,
 'Short': 1,
 'Animation': 2,
 'Comedy': 3,
 'Romance': 4,
 'Sport': 5,
 'News': 6,
 'Drama': 7,
 'Fantasy': 8,
 'Horror': 9,
 'Biography': 10,
 'Music': 11,
 'War': 12,
 'Crime': 13,
 'Western': 14,
 'Family': 15,
 'Adventure': 16,
 'Action': 17,
 'History': 18,
 'Mystery': 19,
 nan: 20,
 'Sci-Fi': 21,
 'Musical': 22,
 'Thriller': 23,
 'Film-Noir': 24,
 'Talk-Show': 25,
 'Game-Show': 26,
 'Reality-TV': 27,
 'Adult': 28}

In [93]:
# make a new column of genre_id
## make new integer genre_id and drop string genres
title_genres['genre_id'] = title_genres['genres_split'].map(genre_map)
title_genres = title_genres.drop(columns='genres_split')

In [102]:
title_genres.head()

tconst  genre_id
0  tt0000001         0
0  tt0000001         1
1  tt0000002         2
1  tt0000002         1
2  tt0000003         2

### Creating "Genre" Table

In [95]:
# make the new dataframe with named columns from the keys and value pairing
genres = pd.DataFrame({'Genre Name': genre_map.keys(),
                            'Genre ID': genre_map.values()})
genres.head()


Genre Name  Genre ID
0  Documentary         0
1        Short         1
2    Animation         2
3       Comedy         3
4      Romance         4

In [96]:
title_basics.head()

tconst titleType            primaryTitle           originalTitle  \
0  tt0000001     short              Carmencita              Carmencita   
0  tt0000001     short              Carmencita              Carmencita   
1  tt0000002     short  Le clown et ses chiens  Le clown et ses chiens   
1  tt0000002     short  Le clown et ses chiens  Le clown et ses chiens   
2  tt0000003     short          Pauvre Pierrot          Pauvre Pierrot   

   isAdult  startYear  endYear  runtimeMinutes                    genres  \
0        0    1894.00      NaN            1.00         Documentary,Short   
0        0    1894.00      NaN            1.00         Documentary,Short   
1        0    1892.00      NaN            5.00           Animation,Short   
1        0    1892.00      NaN            5.00           Animation,Short   
2        0    1892.00      NaN            4.00  Animation,Comedy,Romance   

  genres_split  
0  Documentary  
0        Short  
1    Animation  
1        Short  
2    Animation

In [97]:
title_basics = title_basics.drop(columns=['originalTitle','isAdult','titleType','endYear','genres_split'], axis=1)
title_basics[:7]

tconst            primaryTitle  startYear  runtimeMinutes  \
0  tt0000001              Carmencita    1894.00            1.00   
0  tt0000001              Carmencita    1894.00            1.00   
1  tt0000002  Le clown et ses chiens    1892.00            5.00   
1  tt0000002  Le clown et ses chiens    1892.00            5.00   
2  tt0000003          Pauvre Pierrot    1892.00            4.00   
2  tt0000003          Pauvre Pierrot    1892.00            4.00   
2  tt0000003          Pauvre Pierrot    1892.00            4.00   

                     genres  
0         Documentary,Short  
0         Documentary,Short  
1           Animation,Short  
1           Animation,Short  
2  Animation,Comedy,Romance  
2  Animation,Comedy,Romance  
2  Animation,Comedy,Romance

In [98]:
title_basics = title_basics.drop(columns=['genres'], axis=1)
title_basics.head()

tconst            primaryTitle  startYear  runtimeMinutes
0  tt0000001              Carmencita    1894.00            1.00
0  tt0000001              Carmencita    1894.00            1.00
1  tt0000002  Le clown et ses chiens    1892.00            5.00
1  tt0000002  Le clown et ses chiens    1892.00            5.00
2  tt0000003          Pauvre Pierrot    1892.00            4.00

In [99]:
tmdb_data = tmdb_data.drop(columns=['adult','overview','video'], axis=1)
tmdb_data.head()

imdb_id       budget                                             genres  \
7   tt0118694    150000.00  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
19  tt0120467    120000.00  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   
22  tt0120630  45000000.00  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
25  tt0120753   8000000.00  [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...   
26  tt0120755 125000000.00  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   

         id original_language            original_title  popularity  \
7    843.00                cn                      花樣年華       23.46   
19 19085.00                en                    Vulgar        4.57   
22  7443.00                en               Chicken Run       38.20   
25   318.00                en  The Million Dollar Hotel       11.32   
26   955.00                en    Mission: Impossible II       33.85   

   release_date      revenue  runtime    status                     title  \
7    2000-09-29  12854953.00    99.00  Released      In the Mood for Love   
19   2002-04-26     14904.00    87.00  Released                    Vulgar   
22   2000-06-23 224834564.00    84.00  Released               Chicken Run   
25   2000-02-09    105983.00   122.00  Released  The Million Dollar Hotel   
26   2000-05-24 546388105.00   123.00  Released    Mission: Impossible II   

    vote_average  vote_count certification  
7           8.11     2163.00            PG  
19          5.50       44.00             R  
22          6.70     4313.00             G  
25          5.92      274.00             R  
26          6.11     5706.00         PG-13

In [100]:
tmdb_data = tmdb_data.drop(columns=['id','original_title','original_language','runtime','vote_average','vote_count'], axis=1)
tmdb_data.head()

imdb_id       budget                                             genres  \
7   tt0118694    150000.00  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
19  tt0120467    120000.00  [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...   
22  tt0120630  45000000.00  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
25  tt0120753   8000000.00  [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...   
26  tt0120755 125000000.00  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   

    popularity release_date      revenue    status                     title  \
7        23.46   2000-09-29  12854953.00  Released      In the Mood for Love   
19        4.57   2002-04-26     14904.00  Released                    Vulgar   
22       38.20   2000-06-23 224834564.00  Released               Chicken Run   
25       11.32   2000-02-09    105983.00  Released  The Million Dollar Hotel   
26       33.85   2000-05-24 546388105.00  Released    Mission: Impossible II   

   certification  
7             PG  
19             R  
22             G  
25             R  
26         PG-13

## Moving Data to MySQL

In [36]:
## filling the data types for the title basics table
from sqlalchemy.types import *
# calculating the max string lengths
key_len = title_basics['tconst'].fillna('').map(len).max()
title_len = title_basics['primaryTitle'].fillna('').map(len).max()
# creating a schema dictionary using sqlalchemy datatype objects
df_schema = {
    "tconst": String(key_len+1), 
    "primaryTitle": Text(title_len+1),
    'startYear':Float(),
    'endYear':Float(),
    'runtimeMinutes':Integer()}

In [37]:
from sqlalchemy import create_engine,Table, Column, Integer, String, MetaData, VARCHAR
from sqlalchemy_utils import create_database, database_exists
import pymysql
pymysql.install_as_MySQLdb()

# Additional Imports
import os, json, math, time
from tqdm.notebook import tqdm_notebook
from scipy import stats
import scipy
scipy.__version__

'1.9.3'

In [38]:
## creating a connection string using credentials
username = "root"
password = "Dino2010" # (or whatever password you chose during mysql installation)
db_name = "Movie"
connection = f"mysql+pymysql://{username}:{password}@localhost/{db_name}"

In [39]:
engine = create_engine(connection)
engine

Engine(mysql+pymysql://root:***@localhost/Movie)

In [40]:
# Checking if the database exists with function
if database_exists(connection):
    print('It exists!')
else:
    create_database(connection)
    print('Database created!')

It exists!


In [41]:
# save the table df_filtered to SQL 
df_filtered.to_sql('df_filtered',engine,if_exists='replace',index=False)

402

In [103]:
title_genres.to_sql('title_genres',engine,if_exists='replace',index=False)

15507172

In [44]:
# transferring the other dataframes to mysql
ratings.to_sql('ratings',engine,if_exists='replace',index=False)
genres.to_sql('genres',engine,if_exists='replace',index=False)
title_basics.to_sql('title_basics',engine,if_exists='replace',index=False)


In [45]:
q = """SELECT * FROM title_basics;"""
pd.read_sql(q, engine)

tconst               primaryTitle  startYear  runtimeMinutes
0         tt0000001                 Carmencita    1894.00            1.00
1         tt0000001                 Carmencita    1894.00            1.00
2         tt0000002     Le clown et ses chiens    1892.00            5.00
3         tt0000002     Le clown et ses chiens    1892.00            5.00
4         tt0000003             Pauvre Pierrot    1892.00            4.00
...             ...                        ...        ...             ...
15507167  tt9916852              Episode #3.20    2010.00             NaN
15507168  tt9916856                   The Wind    2015.00           27.00
15507169  tt9916880  Horrid Henry Knows It All    2014.00           10.00
15507170  tt9916880  Horrid Henry Knows It All    2014.00           10.00
15507171  tt9916880  Horrid Henry Knows It All    2014.00           10.00

[15507172 rows x 4 columns]

In [46]:
q = """SELECT * FROM df_filtered;"""
pd.read_sql(q, engine)

imdb_id       budget  \
0    tt0118694    150000.00   
1    tt0120467    120000.00   
2    tt0120630  45000000.00   
3    tt0120753   8000000.00   
4    tt0120755 125000000.00   
..         ...          ...   
397  tt2321658         4.00   
398  tt2457382     39175.00   
399  tt3840036     23700.00   
400  tt3848476     23505.00   
401  tt4346442     20000.00   

                                                genres  popularity  \
0    [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...       23.46   
1    [{'id': 18, 'name': 'Drama'}, {'id': 27, 'name...        4.57   
2    [{'id': 16, 'name': 'Animation'}, {'id': 35, '...       38.20   
3    [{'id': 18, 'name': 'Drama'}, {'id': 53, 'name...       11.32   
4    [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...       33.85   
..                                                 ...         ...   
397                [{'id': 99, 'name': 'Documentary'}]        0.60   
398                     [{'id': 27, 'name': 'Horror'}]        1.03   
399                     [{'id': 27, 'name': 'Horror'}]        1.21   
400                     [{'id': 27, 'name': 'Horror'}]        0.60   
401                     [{'id': 35, 'name': 'Comedy'}]        0.73   

    release_date      revenue    status                     title  \
0     2000-09-29  12854953.00  Released      In the Mood for Love   
1     2002-04-26     14904.00  Released                    Vulgar   
2     2000-06-23 224834564.00  Released               Chicken Run   
3     2000-02-09    105983.00  Released  The Million Dollar Hotel   
4     2000-05-24 546388105.00  Released    Mission: Impossible II   
..           ...          ...       ...                       ...   
397   2008-01-01         4.00  Released   SHAC Action & The Truth   
398   2001-03-15     32124.00  Released                     Dafan   
399   2001-11-09     20540.00  Released            Maut Ki Haveli   
400   2001-06-01     18412.00  Released                  Tahkhana   
401   2001-02-11     50000.00  Released  D'Unbelievables D'Mother   

    certification  
0              PG  
1               R  
2               G  
3               R  
4           PG-13  
..            ...  
397          None  
398          None  
399          None  
400          None  
401          None  

[402 rows x 9 columns]

In [101]:
q = """SELECT * FROM title_genres;"""
pd.read_sql(q, engine)

Empty DataFrame
Columns: [tconst, genre_id]
Index: []

In [49]:
# displaying the 5 tables 
q = """SHOW TABLES;"""
pd.read_sql(q, engine)

Tables_in_movie
0     df_filtered
1          genres
2         ratings
3    title_basics
4    title_genres